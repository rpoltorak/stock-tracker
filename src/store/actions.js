export const ActionTypes = {
  ADD_COMPANY: "ADD_COMPANY",
  REMOVE_COMPANY: "REMOVE_COMPANY",
};
