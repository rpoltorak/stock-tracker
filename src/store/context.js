import React from "react";

export const Store = React.createContext({
  companies: {
    byId: {},
    ids: [],
  },
});
