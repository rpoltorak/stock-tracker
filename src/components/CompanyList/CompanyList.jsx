import React, { useContext } from "react";
import { Link } from "@reach/router";
import { Store } from "../../store";
import { CompanyItem } from "./CompanyItem";

export function CompanyList() {
  const {
    state: { companies },
  } = useContext(Store);

  return companies.ids.length ? (
    companies.ids.map(symbol => (
      <CompanyItem key={symbol} symbol={symbol}></CompanyItem>
    ))
  ) : (
    <div>
      There are no companies yet.{" "}
      <Link to="/new">Track your first company.</Link>
    </div>
  );
}
