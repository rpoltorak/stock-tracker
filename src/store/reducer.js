import omitBy from "lodash.omitby";
import { ActionTypes } from "./actions";

const initialState = {
  companies: {
    byId: {},
    ids: [],
  },
};

export function reducer(state = initialState, action) {
  switch (action.type) {
    case ActionTypes.ADD_COMPANY:
      return addCompany(state, action.payload);
    case ActionTypes.REMOVE_COMPANY:
      return removeCompany(state, action.payload);
    default:
      return state;
  }
}

function addCompany(state, company) {
  const { symbol } = company;

  if (!symbol || state.companies.ids.includes(symbol)) {
    return state;
  }

  return {
    ...state,
    companies: {
      byId: {
        ...state.companies.byId,
        [symbol]: company,
      },
      ids: [...state.companies.ids, symbol],
    },
  };
}

function removeCompany(state, symbol) {
  return {
    companies: {
      byId: omitBy(state.companies.byId, item => item.symbol === symbol),
      ids: state.companies.ids.filter(item => item !== symbol),
    },
  };
}
