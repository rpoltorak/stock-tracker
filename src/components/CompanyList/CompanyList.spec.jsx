import React from "react";
import axios from "axios";
import { render, wait, cleanup } from "@testing-library/react";
import { Store } from "../../store";
import { CompanyList } from "./CompanyList";

jest.mock("axios");

const exampleData = [
  {
    "1. symbol": "GOOGL",
    "2. name": "Alphabet Inc.",
    "3. type": "Equity",
    "4. region": "United States",
    "5. marketOpen": "09:30",
    "6. marketClose": "16:00",
    "7. timezone": "UTC-04",
    "8. currency": "USD",
    "9. matchScore": "0.8889",
  },
];

const setupWithState = () => {
  const state = {
    companies: {
      byId: {
        GOOGL: {
          symbol: "GOOGL.ARG",
          name: "Alphabet Inc.",
          type: "Equity",
          region: "Argentina",
          marketOpen: "09:30",
          marketClose: "6:00",
          timezone: "UTC-04",
          currency: "USD",
          matchScore: "0.8889",
        },
      },
      ids: ["GOOGL"],
    },
  };

  render(
    <Store.Provider value={{ state }}>
      <CompanyList />
    </Store.Provider>,
  );
};

afterEach(() => {
  cleanup();
  axios.mockClear();
});

describe("CompanySearch", () => {
  test("matches the snapshot", async () => {
    axios.mockResolvedValue({ data: { bestMatches: exampleData } });

    const component = setupWithState();

    await wait(() => {
      expect(component).toMatchSnapshot();
    });
  });

  test("renders properly on empty state", () => {
    const state = {
      companies: {
        byId: {},
        ids: [],
      },
    };

    const { getByText } = render(
      <Store.Provider value={{ state }}>
        <CompanyList />
      </Store.Provider>,
    );

    expect(getByText("There are no companies yet.")).toBeTruthy();
  });

  test("makes proper api calls for a list element", async () => {
    setupWithState();

    await wait(() => {
      expect(axios).toHaveBeenCalledTimes(2);
    });
  });
});
