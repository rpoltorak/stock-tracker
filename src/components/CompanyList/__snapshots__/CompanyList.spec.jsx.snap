// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompanySearch matches the snapshot 1`] = `undefined`;
